sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -99,88 +103,105 @@ impl Hyperdebug {
                         config_desc.number(),
                         interface.number()
                     ));
-                // Check the ASCII name of this USB interface.
-                match interface_name.as_str() {
-                    "HyperDebug Shell" => {
+                // Check the class/subclass/protocol of this USB interface.
+                if interface_desc.class_code() == 255
+                    && interface_desc.sub_class_code() == 80
+                    && interface_desc.protocol_code() == 1
+                {
+                    // A serial console interface, use the ascii name to determine if it is the
+                    // HyperDebug Shell, or a UART forwarding interface.
+                    let idx = match interface_desc.description_string_index() {
+                        Some(idx) => idx,
+                        None => continue,
+                    };
+                    let interface_name = match device.read_string_descriptor_ascii(idx) {
+                        Ok(interface_name) => interface_name,
+                        _ => continue,
+                    };
+                    if interface_name.ends_with("Shell") {
                         // We found the "main" control interface of HyperDebug, allowing textual
                         // commands to be sent, to e.g. manipoulate GPIOs.
                         console_tty = Some(Self::find_tty(&interface_path)?)
-                    }
-                    name if name.starts_with("UART") => {
+                    } else {
                         // We found an UART forwarding USB interface.
-                        uart_ttys.insert(name.to_string(), Self::find_tty(&interface_path)?);
+                        uart_ttys.insert(interface_name.to_string(), Self::find_tty(&interface_path)?);
                     }
-                    "SPI" => {
-                        // We found the SPI forwarding USB interface (this one interface allows
-                        // multiplexing physical SPI ports.)
-                        let mut in_endpoint: Option<u8> = None;
-                        let mut out_endpoint: Option<u8> = None;
-                        for endpoint_desc in interface_desc.endpoint_descriptors() {
-                            if endpoint_desc.transfer_type() != rusb::TransferType::Bulk {
-                                continue;
+                }
+                if interface_desc.class_code() == 255
+                    && interface_desc.sub_class_code() == 81
+                    && interface_desc.protocol_code() == 2
+                {
+                    // We found the SPI forwarding USB interface (this one interface allows
+                    // multiplexing physical SPI ports.)
+                    let mut in_endpoint: Option<u8> = None;
```

Comment:
This code seems to be duplicated further down. Factor it out?

Reply:
That is a great idea. Done.